### PR TITLE
[WIP] Enable warnings

### PR DESCRIPTION
**Pull request by bazsi**
(edited)
Trivia:
--enable-extra-warnings is **not** a gcc/clang compiler/linker option. It is a syslog-ng configure option (like any --disable-[some-module] options). It is already part of syslog-ng, marked as recommended to be used, and its purpose was to enable three specific kind of warnings during build.

The ultimate goal is of course that there would not be any kind of warnings: to add -Wall compile option, which is part of this patch. The problem is a lot of warnings appear this way (tens of megabytes of log), which is not practical to handle. So we need to reduce the appearing warnings to a manageable size by whitelisting warnings, and the whitelist would be later reduced in other patches.

In short, this patchset replaces the backlisting method on warnings to whitelisting method, and try to enforce the current whitelisting with travis.
(/edited)

This branch rebases #1128 against current master and refines options somewhat and CLAGS as well.

Major changes against #1128:
 * use g_path_basename() instead of a const basename() implementation
 * it sets the default CFLAGS even if --enable-extra-warnings is not defined, e.g -Wshadow is always defined
 * all warning related CFLAGS are set in the Makefile and not in configure.ac

I am not yet sure if travis should run with --enable-extra-warnings, it bloats the build log. My feeling is that we should merge this without the travis part (e.g. build without the extra warnings even in travis). But I am open to ideas.

It would be great if we could merge this at this time, so we don't need to rebase it again in 6 months time :)

### CODE REVIEW

**Comment by presidento** [comment]
@bazsi, do you want to enable `Wall` and remove the specific warnings? A local machine will generate the warnings (but not Travis). The extra warnings pieces must be evaluated one by one.

**Comment by presidento** [comment]
@bazsi, please merge the trivial fixes, and come up with a new PR with the specific warnings.

**Comment by kira-syslogng** [comment]
Can one of the admins verify this patch?

**Comment by pprindeville** [comment]
Just saw this.  Did I step on this with PR #1398?

**Comment by presidento** [comment]
No, we liked your fixes, this one was too big at the moment for us.

**Comment by pprindeville** [comment]
Okay, good.

**Comment by kira-syslogng** [comment]
This user does not have permission to start the build. Can one of the admins verify this patch and start the build? (admin please type: ok to test)

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by pprindeville** [comment]
Where are the build logs?

**Comment by Kokan** [comment]
@pprindeville 
You could check the `Kira` build log via the *Details* link next to the status, which is going to forward you to a jenkins job (https://ci.syslog-ng.com/jenkins/job/ose-pull-request-test/934/), you could see the result if you log into there with your github account.

You could see the *continuous-integration/travis-ci/pr * that is a Travis CI job.

**Comment by pprindeville** [comment]
Doesn’t work for me... missing Overall Read access.

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: Makefile.am [hunk 1]
```diff
@@ -41,12 +41,62 @@ pkgconfig_DATA		= syslog-ng.pc
...
+	-Wshadow \
+	-fno-common
+
 if ENABLE_EXTRA_WARNINGS
+
 AM_CFLAGS += \
-	-Wshadow	\
-	-Wno-initializer-overrides \
-	-Wno-error=deprecated-declarations
+	-Wno-empty-body \
```
Comment (pprindeville) [inline Makefile.am]: Are all of those warnings really being generated?

File: .travis.yml [hunk 2]
```diff
@@ -44,7 +44,13 @@ before_script:
   - sudo dpkg-reconfigure --frontend noninteractive tzdata
   - ./autogen.sh
   - unset PYTHON_CFLAGS # HACK
-  - CONFIGURE_FLAGS="
+  - if [ "$CC" = "gcc" ]; then
```
Comment (pprindeville) [inline .travis.yml]: Does the indent need to change?

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

File: tests/unit/test_persist_state.c [hunk 3]
```diff
@@ -77,7 +77,7 @@ Test(persist_state, test_persist_state_open_success_on_invalid_file)
   unlink(persist_file);
 
   fd = open(persist_file, O_CREAT | O_RDWR, 0777);
-  write(fd, "aaa", 3);
+  cr_assert(write(fd, "aaa", 3) == 3, "Write error on invalid persist file");
```
Comment (pprindeville) [inline tests/unit/test_persist_state.c]: I'd rather you capture the return value into a variable, and then `cr_assert()` on that.

Also makes it easier to set watchpoints on the variable in gdb, etc.

File: tests/unit/test_persist_state.c [hunk 4]
```diff
@@ -94,7 +94,7 @@ Test(persist_state, test_persist_state_open_fails_on_invalid_file_with_dump)
...
```
Comment (pprindeville) [inline tests/unit/test_persist_state.c]: Ditto

File: modules/stardate/tests/test_stardate.c [hunk 5]
```diff
@@ -38,9 +38,9 @@ stardate_assert(const gchar *msg_str, const int precision, const gchar *expected
 
   char *template_command;
   if (precision == -1)
-    asprintf(&template_command, "$(stardate $UNIXTIME)");
+    g_assert( asprintf(&template_command, "$(stardate $UNIXTIME)") > -1);
```
Comment (pprindeville) [inline modules/stardate/tests/test_stardate.c]: Please capture the return value into a variable and then `g_assert()` on that, instead.

Also, `>= 0` is perhaps a little more intuitive than `> -1`.

File: modules/stardate/tests/test_stardate.c [hunk 6]
```diff
@@ -38,9 +38,9 @@ stardate_assert(const gchar *msg_str, const int precision, const gchar *expected
...
   else
-    asprintf(&template_command, "$(stardate --digits %d $UNIXTIME)", precision);
+    g_assert( asprintf(&template_command, "$(stardate --digits %d $UNIXTIME)", precision) > -1);
```
Comment (pprindeville) [inline modules/stardate/tests/test_stardate.c]: As above.

File: modules/stardate/tests/test_stardate.c (hunk 5 above)
Comment (szemere) [reply modules/stardate/tests/test_stardate.c]: Thank you for notice, nice catch! It is okay for you if I just assert for the error case (http://man7.org/linux/man-pages/man3/asprintf.3.html) to avoid segfault? The actual return value is tested in the next line.

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by furiel** [comment]
I found an extra warning in the code: that is somewhat hidden:
sockaddr_to_dnscache_key -Werror=return-type in case IPV6 is disabled. If it does not mean much extra work, it would be great to check the warnings that are generated when
[...]configure [...] --disable-all-modules option is used.

**Comment by kira-syslogng** [comment]
Build FAILURE, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

File: Makefile.am [hunk 7]
```diff
@@ -41,12 +41,62 @@ pkgconfig_DATA		= syslog-ng.pc
...
+	-Wshadow \
+	-fno-common
+
 if ENABLE_EXTRA_WARNINGS
+
 AM_CFLAGS += \
-	-Wshadow	\
-	-Wno-initializer-overrides \
-	-Wno-error=deprecated-declarations
+        -Wno-empty-body \
```
Comment (furiel) [inline Makefile.am]: This patch confuses me a lot about the goal of this PR. Before seeing this, I was thinking the goal is to switch as much warnings to errors, as possible, otherwise we silence the errors. Now this patch tries to works the opposite way: downgrading errors to warnings.

Seeing this patch, silencing all warnings would not be less appealing. Lets discuss about this patch irl

File: Makefile.am [hunk 8]
```diff
@@ -41,12 +41,62 @@ pkgconfig_DATA		= syslog-ng.pc
 
 AM_CPPFLAGS		= -I$(top_srcdir)/lib -I$(top_srcdir)/modules -I$(top_builddir)/lib -I$(top_builddir)/modules
 
-AM_CFLAGS =
+
+# configure script gives us -g and/or -O2, but all warning related settings
+# should be here
+
+AM_CFLAGS = -Wall \
```
Comment (furiel) [inline Makefile.am]: It would be great if we could ensure the feature parity with cmake.

File: libtest/testutils.h [hunk 9]
```diff
@@ -142,7 +141,7 @@ gchar *basename_const(const gchar *file);
       current_testcase_description = g_string_sized_new(0); \
       g_string_printf(current_testcase_description, description_template, ##__VA_ARGS__); \
       current_testcase_function = __FUNCTION__; \
-      current_testcase_file = basename_const(__FILE__); \
```
Comment (furiel) [inline libtest/testutils.h]: Please roll down this patch to the commit that introduces basename_const

File: syslog-ng/main.c [hunk 10]
```diff
@@ -166,6 +167,8 @@ version(void)
          ON_OFF_STR(SYSLOG_NG_ENABLE_TCP_WRAPPER),
          ON_OFF_STR(SYSLOG_NG_ENABLE_LINUX_CAPS),
          ON_OFF_STR(SYSLOG_NG_ENABLE_SYSTEMD));
+
+  g_free(installer_version);
```
Comment (furiel) [inline syslog-ng/main.c]: Please merge this with the patch that introduces the g_strdup of installer_version.

File: tests/unit/test_persist_state.c [hunk 11]
```diff
@@ -77,7 +77,8 @@ Test(persist_state, test_persist_state_open_success_on_invalid_file)
   unlink(persist_file);
 
   fd = open(persist_file, O_CREAT | O_RDWR, 0777);
-  write(fd, "aaa", 3);
+  ssize_t ret = write(fd, "aaa", 3);
+  cr_assert_eq(ret, 3, "Write error on invalid persist file");
```
Comment (furiel) [inline tests/unit/test_persist_state.c]: Very optional: if we add some error message, can we emit errno as well? Same for the other occurrence below.

File: lib/eventlog/src/evtctx.c [hunk 12]
```diff
@@ -94,12 +94,13 @@ evt_read_config(EVTCONTEXT *ctx)
...
   char line[1024];
+  char *s;
 
   fp = fopen(__PATH_ETC_EVENTLOG_CONF, "r");
   if (!fp)
     return;
-  fgets(line, sizeof(line), fp);
-  while (!feof(fp))
+  s = fgets(line, sizeof(line), fp);
+  while ((s != NULL) && !feof(fp))
```
Comment (furiel) [inline lib/eventlog/src/evtctx.c]: Do we want to handle the error as well?

File: modules/stardate/tests/test_stardate.c [hunk 13]
```diff
@@ -37,10 +37,13 @@ stardate_assert(const gchar *msg_str, const int precision, const gchar *expected
...
   char *template_command;
+  int ret_val;
   if (precision == -1)
-    asprintf(&template_command, "$(stardate $UNIXTIME)");
+    ret_val = asprintf(&template_command, "$(stardate $UNIXTIME)");
   else
-    asprintf(&template_command, "$(stardate --digits %d $UNIXTIME)", precision);
+    ret_val = asprintf(&template_command, "$(stardate --digits %d $UNIXTIME)", precision);
+
+  assert_false(ret_val == -1, "Memory allocation failed in asprintf.");
```
Comment (furiel) [inline modules/stardate/tests/test_stardate.c]: A lot is happening in this patch. I suggest a reword at least, the fixing indentation as a summary is misleading. Maybe worth splitting this patch?

File: lib/template/tests/test_template_compile.c [hunk 14]
```diff
@@ -59,17 +59,28 @@ Plugin hello_plugin = TEMPLATE_FUNCTION_PLUGIN(hello, "hello");
     element.type;  \
     element.msg_ref; }
 
-#define assert_compiled_template(text, default_value, spec, type, msg_ref) \
+#define assert_compiled_template(set_text, set_default_value, spec, type, msg_ref) \
   do { \
+    const gchar *text; \
+    set_text; \
```
[the remark this answers is not in the document]
Comment (furiel) [reply lib/template/tests/test_template_compile.c]: Ok I think I understand how this works... As we already call like `assert_compiled_template(text = "", ...);`, these lines will expand into

`const gchar *text;
text = (text = "");
gchar *text_mut = g_strdup(text);`

while relying that the return value of the assignment carries the value of the assignment.

It was confusing for me that we needed to use exactly `text` and `default_value` in the left side of the assignment, but this restriction is hidden underneath a few macro expansion.  But this patch clarifies this a little bit: the input is not a variable but an expression that needs to be evaluated. @szemere you can ignore my comments, this change is ok I think.

**Comment by kira-syslogng** [comment]
Build SUCCESS, the tests were executed on test branch: master and test suite: functions

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

File: syslog-ng/main.c (hunk 10 above)
Comment (furiel) [reply syslog-ng/main.c]: This request was cosmetic and fixed, though github cannot detect the fix. I delete the whole comment section not to confuse others.

File: modules/stardate/tests/test_stardate.c (hunk 13 above)
Comment (furiel) [reply modules/stardate/tests/test_stardate.c]: This again was cosmetic and fixed, though github cannot detect the fix. I remove the comments so not to confuse others.

**Comment by kira-syslogng** [comment]
success

**Comment by furiel** [comment]
Trivia:
--enable-extra-warnings is **not** a gcc/clang compiler/linker option. It is a syslog-ng configure option (like any --disable-[some-module] options). It is already part of syslog-ng, marked as recommended to be used, and its purpose was to enable three specific kind of warnings during build.

The ultimate goal is of course that there would not be any kind of warnings: to add -Wall compile option, which is part of this patch. The problem is a lot of warnings appear this way (tens of megabytes of log), which is not practical to handle. So we need to reduce the appearing warnings to a manageable size by whitelisting warnings, and the whitelist would be later reduced in other patches.

In short, this patchset replaces the backlisting method on warnings to whitelisting method, and try to enforce the current whitelisting with travis.

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by furiel** [comment]
I was thinking about this PR, and I am starting to feel this is not a good way that we try to approach the issue.

I created a statistics about the warnings we are talking about. If someone want to experiment, here is what I did
```
$ ../syslog-ng/configure --enable-extra-warnings --enable-debug --disable-java --prefix=`pwd`/../root
$ make -j 2> stats.txt
```

```
$ grep -F "[-W" stats.txt | grep -v ivykis | awk '{ print $NF}' | sort | wc -l
8589
$ grep -F "[-W" stats.txt | grep -v ivykis | awk '{ print $NF}' | sort | uniq -c
     26 [-Waggregate-return]
     16 [-Wbad-function-cast]
    174 [-Wcast-qual]
   1685 [-Wconversion]
     52 [-Wdiscarded-qualifiers]
      5 [-Wdouble-promotion]
     16 [-Wfloat-conversion]
      4 [-Wfloat-equal]
      8 [-Wformat-nonliteral]
      2 [-Wimplicit-function-declaration]
    370 [-Wmissing-prototypes]
      1 [-Wnested-externs]
      2 [-Wold-style-declaration]
     16 [-Wpointer-arith]
     34 [-Wpointer-sign]
    356 [-Wredundant-decls]
   1014 [-Wsign-compare]
   4715 [-Wsign-conversion]
     16 [-Wstrict-prototypes]
     10 [-Wsuggest-attribute=format]
      2 [-Wsuggest-attribute=noreturn]
     54 [-Wswitch-default]
      1 [-Wtype-limits]
     10 [-Wundef]
````

This means we have 8589 warnings in total. This is simply too much to much to emit. It will not help developers, because it is impossible to spot changes in such mass. But if we start to filter warnings out, then our original goal breaks: to put pressure to eliminate old warnings and avoid creating new ones.

The way I would approach is we create an issue with the statistics included, and fix warning types one by one. Each time a warning type is completely eliminated, we put an enforcing rule into travis, so to maintain quality. I would would cleanup this pr, because there are still a few good ideas/patches to salvage, like the clarification how we handle warning related compiler options, and the part when explicitely disable some warnings like
```
	-Wno-missing-field-initializers \
	-Wno-stack-protector \
	-Wno-unused-parameter \
	-Wno-variadic-macros
```
but drop the rest.

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by kira-syslogng** [comment]
success

**Comment by furiel** [comment]
Thank you all for your help. Everything is merged from this patch that we needed for now. The remaining warnings are put into issues. @gaborznagy 's -Wold-style-definition is neither forgotten, it is put into #1965 

So the plan is each warnings will be handled with issues and separate PRs. Conceptually, this PR is split to subtask. We can close this PR now.

**Comment by szemere** [comment]
@furiel Thank You for picking up this thread. I also would like to appreciate all the efforts everyone else put into this topic. @pprindeville Thank You for your review notes.
I think this work will highly increase the code quality of syslog-ng, and the enabled warnings will prevent many accidents in the future.
